base_layer/core/src/proof_of_work/siphash.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021 The Grin Developers
+// Modifications copyright 2025 The Tari Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// This file was modified from its original form:
+// https://github.com/mimblewimble/grin/blob/master/core/src/pow/siphash.rs
+
+//! Simple implementation of the siphash 2-4 hashing function from
+//! Jean-Philippe Aumasson and Daniel J. Bernstein.
+
+// Parameters to the siphash block algorithm. Used by Cuckaroo but can be
+// seen as a generic way to derive a hash within a block of them.
+const SIPHASH_BLOCK_BITS: u64 = 6;
+const SIPHASH_BLOCK_SIZE: u64 = 1 << SIPHASH_BLOCK_BITS;
+const SIPHASH_BLOCK_MASK: u64 = SIPHASH_BLOCK_SIZE - 1;
+
+// helper macro for left rotation
+macro_rules! rotl {
+    ($num:expr, $shift:expr) => {
+        $num = ($num << $shift) | ($num >> (64 - $shift));
+    };
+}
+
+/// Utility function to compute a single siphash 2-4 based on a seed and
+/// a nonce
+pub fn siphash24(v: &[u64; 4], nonce: u64) -> u64 {
+    let mut siphash = SipHash24::new(v);
+    siphash.hash(nonce, 21); // 21 is standard rotation constant
+    siphash.digest()
+}
+
+/// Builds a block of siphash values by repeatedly hashing from the nonce
+/// truncated to its closest block start, up to the end of the block. Returns
+/// the resulting hash at the nonce's position.
+pub fn siphash_block(v: &[u64; 4], nonce: u64, rot_e: u8, xor_all: bool) -> u64 {
+    // beginning of the block of hashes
+    let nonce0 = nonce & !SIPHASH_BLOCK_MASK;
+    let nonce_i = nonce & SIPHASH_BLOCK_MASK;
+    let mut nonce_hash = vec![0u64; SIPHASH_BLOCK_SIZE as usize];
```

Comment:
Nit: This could be slightly faster on the stack
```suggestion
    let mut nonce_hash = [0u64; SIPHASH_BLOCK_SIZE as usize];
```

Reply:
I've left this file the same as the grin